### ISSUE TITLE
Performance Regression in Pydantic V2 when Validating Large Number of SQLAlchemy Models

### ISSUE THREAD

**Issue by MprBol** [pydantic/pydantic#7993]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

## Issue Description
We're experiencing a performance issue in our closed-source project, ProjectA, after upgrading to Pydantic V2. The 'search' API call that is expected to return and serialize up to 4000 SQLAlchemy models is exhibiting a significant slowdown.

### Expected Behavior
With Pydantic V1, the API call would complete in about 1.2 seconds (nb; by using orjson for serialisation, but the described slowdown doesnt seem related to that). 

### Current Behavior
The same process with Pydantic V2 takes about 40-60 seconds.

### Steps to Reproduce
Due to ProjectA being closed source, I'm unable to provide a reproduction scenario. However, I have narrowed down the performance bottleneck to the conversion of SQLAlchemy models to Pydantic models using the profiling method described below.

### Profiling Information
Profiling within `routers/findings.py::search()` function is as follows:

```python
record_timing(
    request,
    appname="findings.search.before.pydantic",
    note="After database query before pydanticalization",
)
fmschemas = [schemas.Finding.model_validate(bla) for bla in fms]
record_timing(
    request,
    appname="findings.search.after.pydantic.before.json",
    note="After database query, after pydanticalization before json",
)
```

The timing for `findings.search.after.pydantic.before.json` consistently shows 40 to 60 seconds with about ~1400 finding results to be serialized. 

### Model Definition
Our `FindingModel` is defined with multiple inheritance and is quite complex:

```python
class BaseFinding(BaseSchema):
    scan_identifier: str
    name: str
    alias: Optional[str] = None
    type: str
    description: str = ""
    cvss: float = 0  # prepare default value for eventual removal, keeping it here for backwards compatibility
    severity: Severity = Severity.Unknown
    severity_override: Optional[Severity] = None
    severity_rescaled: Optional[Severity] = None
    severity_original: Optional[Severity] = None
    rescaled_ruleset_md5: Optional[str] = None
    cvss_rescaled: Optional[float] = None
    cvss3_original: Optional[str] = None
    cvss3_rescaled: Optional[str] = None
    status: FindingStatus = FindingStatus.New
    status_original: Optional[FindingStatus] = None
    status_override: Optional[FindingStatus] = None
    status_rescaled: Optional[FindingStatus] = None
    port: int
    data: Dict = {}
    meta_data: Dict = {}
    duplicate_of_id: Optional[int] = None
    location_id: int

class FindingFlat(BaseFinding):
    id: int
    date_created: datetime.datetime
    date_updated: datetime.datetime
    date_last_reopened: datetime.datetime
    date_last_rawresult: Optional[datetime.datetime] = None
    date_last_rescaled: Optional[datetime.datetime] = None
    has_duplicates: bool

class Finding(FindingFlat):
    rawresults: Optional[List[RawResultFlat]] = None
    actions: Optional[List[ActionFlat]] = None
    location: Optional["LocationFlat"] = None
    duplicate_of: Optional[FindingFlat] = None
    duplicates: Optional[List[FindingFlat]] = None
    jiratickets: Optional[List[JiraTicketFlat]] = None
```

### Additional Context
The `FindingModel` uses multiple inheritance and is expected to handle a significant number of instances. We suspect the validation process of SQLAlchemy models to Pydantic models is the performance bottleneck. When  using `model_construct()` the API query is a lot quicker (~1.2 seconds), however it seems model_construct cannot read the attributes from sqlalchemy model as all fields are empty/left to default. The models are coming from the db, and are therefore trusted (and should already vallidated by pydantic before the data came in the database). However we probably do need pydantic's logic of parsing the other types (like eg Location) so I dont think that's a good alternative to model_validate(). 

As ProjectA is closed source, reproducing this issue may not be feasible externally. I'm seeking insights into whether this is a recognized issue with Pydantic V2 and if there are any known workarounds or optimizations that can be applied.

Thank you for looking into this.

### Python, Pydantic & OS Version

```Text
pydantic version: 2.4.2
        pydantic-core version: 2.10.1
          pydantic-core build: profile=release pgo=false
                 install path: /Users/mpronk/gits/shield/shieldba/.venv/lib/python3.10/site-packages/pydantic
               python version: 3.10.12 (main, Jun  7 2023, 00:45:35) [Clang 14.0.3 (clang-1403.0.22.14.1)]
                     platform: macOS-13.6-x86_64-i386-64bit
             related packages: mypy-1.6.1 fastapi-0.103.0 typing_extensions-4.8.0 pydantic-settings-2.0.3


- SQLAlchemy Version: 1.4.50
```

**Comment by MprBol** [comment]
This issue can be closed! The problem was in my solution of a custom replacement for GetterDict, which I used to read SQLAlchemy models and provide the models' default value if the loading of that column is set to RaiseLoad. I do this to prevent having to create many Pydantic Models while still keeping the performance gains of not loading all fields from the database all the time. 

The logic I wrote to fetch the default value from the pydantic model was not cached, so it was rebuilding the pydantic model for each missed attribute. This caused a huge performance penalty. I fixed this by adding caching.

TLDR, the working code is:
```python
class SqlAlchemyGetterDict:
    _defaults_cache: Dict[str, Dict[str, Any]] = {}

    def __init__(self, obj: Any, schema_cls: Type["BaseSchema"]):
        self._obj = obj
        self._schema_cls = schema_cls
        # Initialize the cache for this schema class if it doesn't exist yet.
        if schema_cls.__name__ not in self._defaults_cache:
            self._defaults_cache[schema_cls.__name__] = {}

    def __getattr__(self, key: str) -> Any:
        schema_name = self._schema_cls.__name__

        try:
            return getattr(self._obj, key)
        except InvalidRequestError as ex:
            if key in self._defaults_cache[schema_name]:
                # Return the cached default value
                return self._defaults_cache[schema_name][key]
            # Attempt to get the default from the Pydantic model if an InvalidRequestError is caught
            try:
                default_value = self._schema_cls.model_json_schema()["properties"][key][
                    "default"
                ]
                self._defaults_cache[schema_name][key] = default_value
                return default_value
            except KeyError as ex:
                print(f"No default value found for '{key}' in '{schema_name}'")
                return None
            except Exception as ex:
                print(f"Error retrieving default value for '{key}' in '{schema_name}'")
                return None
```

which can be used as follows:

```
class BaseSchema(BaseModel):
    model_config = ConfigDict(
        from_attributes=True, str_strip_whitespace=True, extra="forbid"
    )

    @root_validator(pre=True)
    def run_root_validator(cls, values):
        if type(values) == dict:
            return values

        values = SqlAlchemyGetterDict(values, cls)

        return values
```

With the above logic; you can do
```
fm = sqlalchemy.FindingModel().without('data').options(raiseload=true).get() # pseudocode to fetch sqlalchemy model from db
f = Finding.model.validate(fm)
f.dict()
```
